File: elixir/pascals-triangle/pascals_triangle.exs
defmodule PascalsTriangle do
  @doc """
  Calculates the rows of a pascal triangle
  with the given height
  """
  @spec rows(integer) :: [[integer]]
  def rows(num) do
    (0 .. num - 1) |> Enum.map(&build_row/1)
  end

  defp build_row(row) do
    (0 .. row) |> Enum.map(fn col -> cell(row, col) end)
  end

  defp cell(_, 0), do: 1
  defp cell(x, x), do: 1
  defp cell(row, col), do: cell(row - 1, col) + cell(row - 1, col - 1)

end

File: scala/pascals-triangle/src/main/scala/PascalsTriangle.scala
object PascalsTriangle {

  def triangle(rows: Int): List[List[Int]] = {
    require(rows > 0, "rows must be greater than 0")
    (0 until rows).toList map buildRow
  }

  private def buildRow(row: Int): List[Int] = {
    (0 to row).toList map(cell(row, _))
  }

  private def cell(row: Int, col: Int): Int = {
    (factorial(row) / (factorial(col) * factorial(row - col))).toInt
  }

  private def factorial(n: Long): Long = {
    (1L to n).foldLeft(1L) ((x, y) => x * y)
  }

}
